src/database.cc

```diff
@@ -356,6 +361,131 @@ NAN_METHOD(Database::Configure) {
     NanReturnValue(args.This());
 }
 
+NAN_METHOD(Database::RegisterFunction) {
+    NanScope();
+    Database* db = ObjectWrap::Unwrap<Database>(args.This());
+
+    REQUIRE_ARGUMENTS(2);
+    REQUIRE_ARGUMENT_STRING(0, functionName);
+    REQUIRE_ARGUMENT_FUNCTION(1, callback);
+
+    FunctionBaton *baton = new FunctionBaton(db, *functionName, callback);
+    sqlite3_create_function(
+        db->_handle,
+        *functionName,
+        -1, // arbitrary number of args
+        SQLITE_UTF8 | SQLITE_DETERMINISTIC,
+        baton,
+        FunctionBeginExecute,
+        NULL,
+        NULL);
+
+    NanReturnValue(args.This());
+}
+
+void Database::FunctionBeginExecute(sqlite3_context *context, int argc, sqlite3_value **argv) {
+    FunctionBaton *fn = (FunctionBaton *)sqlite3_user_data(context);
+    ExecuteFunctionBaton *baton = new ExecuteFunctionBaton(fn, context, argc, argv);
+
+    // call to the main thread (uv_default_loop) for JS function to execute &
+    // wait for the result
+    uv_mutex_t mutex;
+    uv_mutex_init(&mutex);
+    uv_cond_init(&baton->condition);
+    uv_async_init(uv_default_loop(), &baton->async, (uv_async_cb)Database::AsyncFunctionExecute);
+    uv_async_send(&baton->async);
+    uv_mutex_lock(&mutex);
+    uv_cond_wait(&baton->condition, &mutex);
```

Comment:
`uv_cond_wait` may produce [spurious wakeups](http://docs.libuv.org/en/latest/threading.html#conditions). We need to check whether the function call is actually complete.